packages/host/app/resources/metamask.ts

```
@@ -0,0 +1,95 @@
+import { tracked } from '@glimmer/tracking';
+import { Resource } from 'ember-resources/core';
+import { enqueueTask, restartableTask } from 'ember-concurrency';
+
+// window is a global object that is available in this resource
+declare global {
+  interface Window {
+    ethereum: any;
+  }
+}
+
+class MetaMaskResource extends Resource {
+  @tracked connected = false;
+  @tracked chainId = -1; // the chain id of the metamask connection (not the card)
+
+  setup() {
+    this.doInitialize.perform();
+    if (this.isMetamaskInstalled()) {
+      window.ethereum.on('chainChanged', this.handleChainChanged.bind(this));
+    }
+    return this;
+  }
+
+  private doInitialize = enqueueTask(async () => {
+    if (this.isMetamaskInstalled()) {
+      let chainId = this.getChainId();
+      let connected = await this.isMetamaskConnected();
+      this.chainId = chainId;
+      this.connected = connected;
+    }
+  });
+
+  teardown() {
+    this.doInitialize.cancelAll();
+  }
+
+  isMetamaskInstalled() {
+    return typeof window.ethereum !== 'undefined';
+  }
+
+  async isMetamaskConnected() {
+    try {
+      if (!this.isMetamaskInstalled()) {
+        return false;
+      }
+      let accounts = await window.ethereum.request({
+        method: 'eth_requestAccounts',
+      });
+      return accounts.length > 0;
+    } catch (e) {
+      return false;
```

I think it would be good to not swallow the error here but to track the error reason as an error state so that we can display it

Similar code: https://github.com/cardstack/cardstack/blob/main/packages/safe-tools-client/app/utils/chain-connection-manager.ts#L391

---

Resolved https://github.com/cardstack/boxel/pull/372/commits/742fa15685233e1d9dcd87a64e5a91cae4053775